### PR TITLE
Expose shadow texture size for directional lighting in SDF

### PR DESCRIPTION
**Pull request by athenaz2**
# 🎉 New feature

Related to https://github.com/gazebosim/gz-rendering/pull/1034, view that for details.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by athenaz2** [comment]
Only recently realized that my PR breaks ABI, so PR is for `main`. I've been working with the default gz-gui, gz-rendering branches as the base until now. Will test my changes (to both gz-gui, gz-rendering) with `main` branch asap.

**Comment by jennuine** [comment]
@iche033 if you have some bandwidth, would you be able to review this since you already have some context with https://github.com/gazebosim/gz-rendering/pull/1034?

**Comment by iche033** [comment]
> @iche033 if you have some bandwidth, would you be able to review this since you already have some context with [gazebosim/gz-rendering#1034](https://github.com/gazebosim/gz-rendering/pull/1034)?

yep will do

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 1]
```diff
@@ -1464,6 +1486,42 @@ void MinimalScene::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
...
+                rendering::LightType::DIRECTIONAL, texSize))
+            {
+              gzerr << "Unable to set <texture_size> to '" << texSizeStr.str()
+                    << "' using default texture size" << std::endl;
+            }
+          }
+          else
+          {
+            gzerr << "Light type [" << lightType << "] is not supported."
+                  << std::endl;
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.cc:1519]:
```suggestion
            gzerr << "Setting shadow <texture_size> for light type [" 
                  << lightType << "] is not supported." << std::endl;
```

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 2]
```diff
@@ -762,6 +762,14 @@ std::string GzRenderer::Initialize(RenderThreadRhi &_rhi)
     scene->SetSkyEnabled(true);
   }
 
+  if (!scene->SetShadowTextureSize(rendering::LightType::DIRECTIONAL,
+      this->directionalLightTextureSize))
+  {
+    gzerr << "Unable to set <texture_size> to '"
+          << this->directionalLightTextureSize
+          << "' using default texture size" << std::endl;
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.cc:770]:
```suggestion
    gzerr << "Unable to set directional light shadow <texture_size> to '"
          << this->directionalLightTextureSize
          << "'. Using default texture size of "
          << scene->ShadowTextureSize(rendering::LightType::DIRECTIONAL)
          << std::endl;
```

File: src/plugins/minimal_scene/MinimalScene.hh [hunk 3]
```diff
@@ -238,6 +239,8 @@ namespace gz::gui::plugins
     /// \brief True if sky is enabled;
     public: bool skyEnable = false;
 
+    public: unsigned int directionalLightTextureSize = 2048u;
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.hh:242]: add doxygen comment

File: src/plugins/minimal_scene/MinimalScene.hh [hunk 4]
```diff
@@ -368,6 +371,12 @@ namespace gz::gui::plugins
     /// \param[in] _sky True to enable the sky, false otherwise.
     public: void SetSkyEnabled(const bool &_sky);
 
+    /// @brief  \brief Set the shadow texture size for the given light type.
+    /// @param _lightType Light type that creates the shadow
+    /// @param _textureSize Shadow texture size
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.hh:376]: nit: replace `@` with `\` for consistency

File: src/plugins/minimal_scene/MinimalScene.hh [hunk 5]
```diff
@@ -368,6 +371,12 @@ namespace gz::gui::plugins
     /// \param[in] _sky True to enable the sky, false otherwise.
     public: void SetSkyEnabled(const bool &_sky);
 
+    /// @brief  \brief Set the shadow texture size for the given light type.
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.hh:374]:
```suggestion
    /// \brief Set the shadow texture size for the given light type.
```

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 6]
```diff
@@ -1464,6 +1486,42 @@ void MinimalScene::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
...
+      auto textureSizeElem = elem->FirstChildElement("texture_size");
+      if (nullptr != elem && nullptr != textureSizeElem->GetText())
+      {
+        unsigned int texSize;
+        std::stringstream texSizeStr;
+        texSizeStr << std::string(textureSizeElem->GetText());
+        texSizeStr >> texSize;
+        if (texSizeStr.fail())
+        {
+          gzerr << "Unable to set <texture_size> to '" << texSizeStr.str()
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.cc:1501]:
```suggestion
          gzerr << "Unable to set shadow <texture_size> to '" << texSizeStr.str()
```

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 7]
```diff
@@ -1464,6 +1486,42 @@ void MinimalScene::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
...
+        }
+        else
+        {
+          std::string lightType = textureSizeElem->Attribute("light_type");
+          if (lightType == "directional")
+          {
+            if (!renderWindow->SetShadowTextureSize(
+                rendering::LightType::DIRECTIONAL, texSize))
+            {
+              gzerr << "Unable to set <texture_size> to '" << texSizeStr.str()
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.cc:1512]:
```suggestion
              gzerr << "Unable to set shadow <texture_size> to '" << texSizeStr.str()
```

**Review by iche033 — COMMENTED** [review]
just some minor comments of error msgs

File: src/plugins/minimal_scene/MinimalScene.hh (hunk 3 above)
Comment (athenaz2) [reply src/plugins/minimal_scene/MinimalScene.hh:243]: Oops. Added here 0bc87f8c8ff8b2023472279cad3d85bf4b447701

File: src/plugins/minimal_scene/MinimalScene.hh (hunk 4 above)
Comment (athenaz2) [reply src/plugins/minimal_scene/MinimalScene.hh]: 0bc87f8c8ff8b2023472279cad3d85bf4b447701

**Review by iche033 — APPROVED** [review]
Looks good to me

Just a note that this works for the 3D scene in the GUI. In the future, we may want to extend this functionality to shadows on the server side for sensors.

**Pull request merged**